File: src/main/java/frc/robot/commands/swerve/WheelRadiusCharacterization.java
package frc.robot.commands.swerve;

import edu.wpi.first.math.MathUtil;
import edu.wpi.first.math.filter.SlewRateLimiter;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.units.Units;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.swerve.Drivebase;
import org.littletonrobotics.junction.Logger;

/** Characterizes wheel radius by spinning in place and comparing yaw to wheel rotation. */
public final class WheelRadiusCharacterization extends Command {
  private final Drivebase drivebase;
  private final Direction direction;
  private final AngularVelocity characterizationSpeed;
  private final SlewRateLimiter omegaLimiter = new SlewRateLimiter(1.0);

  private double[] startWheelPositionsRad = new double[0];
  private double lastGyroYawRads = 0.0;
  private double gyroYawAccumRads = 0.0;

  public WheelRadiusCharacterization(Drivebase drivebase, Direction direction) {
    this(drivebase, direction, Units.RadiansPerSecond.of(1.0));
  }

  public WheelRadiusCharacterization(
      Drivebase drivebase, Direction direction, AngularVelocity characterizationSpeed) {
    this.drivebase = drivebase;
    this.direction = direction;
    this.characterizationSpeed = characterizationSpeed;
    addRequirements(drivebase);
  }

  @Override
  public void initialize() {
    startWheelPositionsRad = drivebase.getWheelRadiusCharacterizationAngles();

    gyroYawAccumRads = 0.0;
    lastGyroYawRads = drivebase.getGyroInputs().yaw.getRadians();
    omegaLimiter.reset(0.0);
  }

  @Override
  public void execute() {
    drivebase.runWheelRadiusCharacterization(characterizationSpeed.in(Units.RadiansPerSecond) * direction.sign);

    double currentYaw = drivebase.getGyroInputs().yaw.getRadians();
    gyroYawAccumRads += MathUtil.angleModulus(currentYaw - lastGyroYawRads);
    lastGyroYawRads = currentYaw;

    double[] wheelPositions = drivebase.getWheelRadiusCharacterizationAngles();
    double sumRad = 0.0;
    for (int i = 0; i < wheelPositions.length; i++) {
      sumRad += Math.abs(wheelPositions[i] - startWheelPositionsRad[i]);
    }
    double averageWheelPositionRad = wheelPositions.length == 0 ? 0.0 : sumRad / wheelPositions.length;

    Logger.recordOutput(
        drivebase.getName() + "/radiusCharacterization/averageWheelPositionRad",
        averageWheelPositionRad);

    if (averageWheelPositionRad > 1e-6) {
      double currentEffectiveWheelRadiusMeters =
          (gyroYawAccumRads * Drivebase.getDrivebaseRadiusMeters()
                  / averageWheelPositionRad)
              * direction.sign;
      Logger.recordOutput(
          drivebase.getName() + "/radiusCharacterization/wheelRadius",
          currentEffectiveWheelRadiusMeters);
    }
  }

  @Override
  public void end(boolean interrupted) {
    drivebase.applyChassisSpeeds(new ChassisSpeeds());
  }

  public enum Direction {
    CLOCKWISE(-1.0),
    COUNTERCLOCKWISE(1.0);

    final double sign;

    Direction(double sign) {
      this.sign = sign;
    }
  }
}

File: src/main/java/frc/robot/commands/swerve/WheelRadiusCharacterization.kt
package frc.robot.commands.swerve

import edu.wpi.first.math.MathUtil
import edu.wpi.first.math.filter.SlewRateLimiter
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.units.Units.*
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.AngularVelocity
import edu.wpi.first.wpilibj2.command.Command
import frc.robot.subsystems.swerve.Drivebase
import lib.math.units.into
import org.littletonrobotics.junction.Logger
import kotlin.math.abs

class WheelRadiusCharacterization(
    private val drivebase: Drivebase,
    private val direction: Direction,
    private val characterizationSpeed: AngularVelocity = RadiansPerSecond.of(1.0)
): Command() {
    private val omegaLimiter = SlewRateLimiter(1.0)
    private var startWheelPositions: List<Angle> = listOf()

    init {
        addRequirements(drivebase)
    }

    private var lastGyroYawRads = 0.0
    private var gyroYawAccumRads = 0.0

    override fun initialize() {
        startWheelPositions = drivebase.wheelRadiusCharacterizationAngles.map { it.copy() }
        gyroYawAccumRads = 0.0

        lastGyroYawRads = drivebase.gyroInputs.yaw.radians

        omegaLimiter.reset(0.0)
    }

    override fun execute() {
        drivebase.runWheelRadiusCharacterization(characterizationSpeed.times(direction.i.toDouble()))

        gyroYawAccumRads += MathUtil.angleModulus(
            drivebase.gyroInputs.yaw.radians - lastGyroYawRads
        )

        lastGyroYawRads = drivebase.gyroInputs.yaw.radians

        val wheelPositions = drivebase.wheelRadiusCharacterizationAngles
        val averageWheelPositionRad  = startWheelPositions.zip(wheelPositions).sumOf { (start, current) ->
            abs(current.into(Radians) - start.into(Radians))
        } / wheelPositions.size

        Logger.recordOutput("${drivebase.name}/radiusCharacterization/averageWheelPositionRad", averageWheelPositionRad)
        

        val currentEffectiveWheelRadiusMeters = ((gyroYawAccumRads * (drivebase.drivebaseRadius into Meters)) / averageWheelPositionRad) * direction.i

        Logger.recordOutput("${drivebase.name}/radiusCharacterization/wheelRadius", currentEffectiveWheelRadiusMeters)


    }

    enum class Direction(val i: Int) {
        CLOCKWISE(-1),
        COUNTERCLOCKWISE(1)
    }

    override fun end(interrupted: Boolean) {
        drivebase.applyChassisSpeeds(ChassisSpeeds())
    }

}